[export] Make RNNs exportable on GPUs

yiming0416 opened the pull request:
Summary: Fixes https://github.com/pytorch/pytorch/issues/155309

`torch.export` fails to export an RNN model on GPU with `cudnn` enabled. This is because RNN module's `flatten_parameters()` method calls `p.data_ptr()` for aliasing detection, causing "Cannot access data pointer of Tensor" errors on GPU.

This PR fixes this issue by disabling cudnn during export time.

Test Plan:
buck2 run mode/dev-nosan caffe2/test:test_export -- -r test_export_lstm_gpu
buck2 run mode/dev-nosan caffe2/test:test_export -- -r test_export_gru_gpu
buck2 run mode/dev-nosan caffe2/test:test_export -- -r test_export_rnn_flatten_parameters

Rollback Plan:

cc @ezyang @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela @chenyang78

@diff-train-skip-merge

tugsbayasgalan commented on torch/_export/serde/serialize.py:
```diff
@@ -353,7 +353,9 @@ def serialize_tensor_meta(t: torch.Tensor) -> TensorMeta:
-        storage_offset=serialize_sym_int(t.storage_offset()),
+        storage_offset=serialize_sym_int(t.storage_offset())
+        if t.device.type == "cpu"
+        else serialize_sym_int(0),
```
Why is this the case?

tugsbayasgalan commented on torch/export/pt2_archive/_package.py:
```diff
@@ -433,7 +433,7 @@ def _package_constants(
             else:
                 # Only save once when tensors share the same storage
-                tensor_storage = constant.untyped_storage()
+                tensor_storage = constant.cpu().untyped_storage()
```
Do we need this?

ezyang commented:
This is a really huge diff and it deserves a much longer PR description than no description as seen her

albanD commented on torch/nn/modules/rnn.py:
```diff
@@ -252,12 +253,21 @@ def flatten_parameters(self) -> None:
+            }
+            if len(unique_storage_refs) != len(self._flat_weights):
+                return
+        except Exception:
```
This is a very weird pattern to use. The new logic should either work or we should know when it doesn't and not try.

albanD replied:
You can add this implementation to expanded weights?
Or skip expanded weights?

yiming0416 replied:
@albanD Makes sense. Added `untyped_storage()` impl for expanded weights and removed `try...except`

albanD commented on torch/utils/weak.py:
```diff
@@ -361,3 +363,39 @@ def __call__(self):
         return out
+
+
+class StorageWeakRef:
```
The reason this doesn't live here is that, contrary to everything else in this file, this does NOT using python weakref module!
Do we actually want to use this vs using python weakref to implement this?
There are most likely a lot of subtle implications of this and I would like to hear more about this choice!

yiming0416 replied:
@albanD After taking a closer look, I think `StorageWeakRef` is different because
1. It works directly with PyTorch's C++ intrusive pointer in `c10::StorageImpl`. Python's weakref can't work with C++ objects directly
2. `StorageWeakRef.cdata` can be serialized and sent between processes. And python's `weakref` usually works within a single process. I think that's why `StorageWeakRef` was in `multiprocessing/reductions.py` originally.
Given this, I am leaning towards not moving `StorageWeakRef` to `utils/weak.py` and keeping it in `multiprocessing/reductions.py`. But we still use it in `rnn.py`. What do you think?

albanD commented on torch/nn/modules/rnn.py:
```diff
@@ -252,11 +253,12 @@ def flatten_parameters(self) -> None:
+        unique_storage_refs = {
+            StorageWeakRef(p.untyped_storage())
+            for p in self._flat_weights  # type: ignore[union-attr]
+            if p is not None
```
p cannot be None given the previous logic?

albanD commented on torch/nn/modules/rnn.py:
```diff
@@ -252,11 +253,12 @@ def flatten_parameters(self) -> None:
-            p.data_ptr()  # type: ignore[union-attr]
-            for p in self._flat_weights
+        unique_storage_refs = {
+            StorageWeakRef(p.untyped_storage())
```
Why move from data_ptr() to untyped_storage() here if the new method is added to expanded_weights either way?

albanD commented on torch/nn/modules/rnn.py:
```diff
@@ -252,11 +253,12 @@ def flatten_parameters(self) -> None:
-            p.data_ptr()  # type: ignore[union-attr]
-            for p in self._flat_weights
+        unique_storage_refs = {
+            StorageWeakRef(p.untyped_storage())
```
Also why do you need a weakref at all here? You can create a set of int or storage. You don't keep this dict alive anyways.

yiming0416 replied:
@albanD yeah that makes sense. changed to directly using `untyped_storage()`

yiming0416 commented on torch/nn/modules/rnn.py:
```diff
@@ -252,11 +252,9 @@ def flatten_parameters(self) -> None:
-            for p in self._flat_weights
-        }
-        if len(unique_data_ptrs) != len(self._flat_weights):
+        unique_storage_refs = {p.untyped_storage() for p in self._flat_weights}  # type: ignore[union-attr]
```
@albanD Understood the concern. But `p` is always from the `_flat_weights` in rnn which would be `torch.Tensor` or subclasses. I guess this somehow narrows the "niche cases"?

Anyways, I'm fine parking it here.

yiming0416 replied:
@albanD  Then how about this? We first check if the `data_ptr` exist for tensors, if it does, we keep the old way of using `p.data_ptr()`. If it doesn't, we use `p.untyped_storage()` instead. 

In this way, we won't break any existing working tensors but also extend the coverage.

yiming0416 committed:
[export] Make RNNs exportable on GPUs (#163245)

Summary:

Completing https://github.com/pytorch/pytorch/pull/155734

Test Plan:
buck2 run mode/dev-nosan caffe2/test:test_export -- -r test_export_lstm_gpu
buck2 run mode/dev-nosan caffe2/test:test_export -- -r test_export_gru_gpu
buck2 run mode/dev-nosan caffe2/test:test_export -- -r test_export_rnn_flatten_parameters_gpu

Reviewed By: tugsbayasgalan

Differential Revision: D82687470

yiming0416 commented:
@pytorchbot merge -f "approved and landed internally"

huydhn commented:
> @huydhn This has been reverted internally. But reverting this PR failed

I have remove the diff version on the PR description to stop the bot from spamming revert.  It should be ok now

yiming0416 commented:
nvm it's successfully reverted. Will create another one to reland
